prow/git/git.go

```diff
@@ -139,24 +147,16 @@ func (c *Client) Clone(repo string) (*Repo, error) {
 	c.lockRepo(repo)
 	defer c.unlockRepo(repo)
 
-	base := c.base
 	user, pass := c.getCredentials()
```

Comment:
Can we write a test which is something along these lines:

```golang
func TestClone(t *testing.T) {
  cases := []struct{
    name string
    client Client // how are fields set in client?
    repo string // what repo do we clone?
    commands [][]string // what commands does this run?
    expected Repo // what repo does it return?
    err bool // did it return and error?
  }
  defer func() {execWithDir = defaultExecWithDir}()

  for _, tc := range cases {
    var actualCommands [][]string
    execWithDir := func(dir, name string, args ...string) {
     actualCommands = append(actualCommands, [dir, name, args...])
    }
    actual, err := tc.client.Clone(tc.repo)
   // check err
   // check actual == expected
  // check commands
}
```

Reply:
sure. the commands would need to be at least {command []string, err return} to test scenarios with git returning error (retries etc).

Reply:
I don't see why. If anything fails we return the error, right? We already track whether the overall script fails. Tracking the exact command that fails seems over specified.

Reply:
This construction is essentially a mock. It would be useful, for testing purposes, to simulate an error of a call. Without it we would only be able to test happy path. For example, I don't see how to verify the retry behavior without returning error from a mocked call.

Reply:
Oh, I see it would just record the calls issued. Still no way to test off-happypath flows.

Reply:
You control the execWithDir implementation, so assuming the retries are internal to client.Clone() then you could do something like:

```golang 
var retries int 
execWithDir := func(dir, name, string, args ...string) error {
  if args == ["whatever", "we", "need", "to", match"] {
    if retries < someConst { retries++; return fmt.Error("fake failure") }
  }
  // log what we called
  return nil
}
```

See how we test flakiness elsewhere: https://github.com/kubernetes/test-infra/pull/11133